dart/dynamics/HierarchicalIK.cpp

```diff
@@ -0,0 +1,809 @@
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "dart/dynamics/DegreeOfFreedom.h"
+#include "dart/dynamics/HierarchicalIK.h"
+#include "dart/dynamics/BodyNode.h"
+#include "dart/dynamics/EndEffector.h"
+#include "dart/dynamics/Skeleton.h"
+
+namespace dart {
+namespace dynamics {
+
+//==============================================================================
+bool HierarchicalIK::solve(bool _resetConfiguration)
+{
+  if(nullptr == mSolver)
+  {
+    dtwarn << "[HierarchicalIK::solve] The Solver for a HierarchicalIK module "
+           << "associated with [" << mSkeleton.lock()->getName() << "] is a "
+           << "nullptr. You must reset the module's Solver before you can use "
+           << "it.\n";
+    return false;
+  }
+
+  if(nullptr == mProblem)
+  {
+    dtwarn << "[HierarchicalIK::solve] The Problem for a HierarchicalIK module "
+           << "associated with [" << mSkeleton.lock()->getName() << "] is a "
+           << "nullptr. You must reset the module's Problem before you can use "
+           << "it.\n";
+    return false;
+  }
+
+  const SkeletonPtr& skel = getSkeleton();
+
+  if(nullptr == skel)
+  {
+    dtwarn << "[HierarchicalIK::solve] Calling a HierarchicalIK module which "
+           << "is associated with a Skeleton that no longer exists.\n";
+    return false;
+  }
+
+  const size_t nDofs = skel->getNumDofs();
+  mProblem->setDimension(nDofs);
+
+  mProblem->setInitialGuess(skel->getPositions());
+
+  Eigen::VectorXd bounds(nDofs);
+  for(size_t i=0; i < nDofs; ++i)
+    bounds[i] = skel->getDof(i)->getPositionLowerLimit();
+  mProblem->setLowerBounds(bounds);
+
+  for(size_t i=0; i < nDofs; ++i)
+    bounds[i] = skel->getDof(i)->getPositionUpperLimit();
+  mProblem->setUpperBounds(bounds);
+
+  refreshIKHierarchy();
+
+  if(!_resetConfiguration)
+    return mSolver->solve();
```

Comment:
It seems the optimal solution is not applied to the target skeleton after the problem is solved. Currently, the last configuration, which was used to calculate objectives or constraints (or their gradients), is applied. The problem is that it is not guaranteed that the last configuration is the optimal solution.

I could simply verified this fact by comparing the optimal solution, that is stored in the Problem, and the actual configuration of the system as:
```cpp
if(!_resetConfiguration)
  {
    bool wasSolved = mSolver->solve();
    Eigen::VectorXd optSol = mProblem->getOptimalSolution();
    std::cout << (skel->getPositions() - optSol).eval().transpose() << std::endl;
    return wasSolved;
  }
```
I observed that the difference was not zero, which needs to be zero if we correctly apply the optimal solution to the skeleton.

This code should be something like:
```cpp
  if(!_resetConfiguration)
  {
    const bool wasSolved = mSolver->solve();
    setConfiguration(mProblem->getOptimalSolution());
    return wasSolved;
  }
```

Also, this should be applied to `InverseKinematics::solve()` too.